lib/PromoteStackPointer/ComputeStackAccessesBoundsPass.cpp

```diff
@@ -2,58 +2,155 @@
 // This file is distributed under the MIT License. See LICENSE.md for details.
 //
 
+#include <optional>
+#include <utility>
+
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/SmallVector.h"
 #include "llvm/Analysis/LazyValueInfo.h"
+#include "llvm/Analysis/LoopInfo.h"
+#include "llvm/Analysis/PostDominators.h"
+#include "llvm/Analysis/ScalarEvolution.h"
+#include "llvm/Analysis/ScalarEvolutionExpressions.h"
+#include "llvm/IR/BasicBlock.h"
 #include "llvm/IR/ConstantRange.h"
+#include "llvm/IR/Dominators.h"
 #include "llvm/IR/Instructions.h"
 
 #include "revng/Model/IRHelpers.h"
+#include "revng/Support/IRHelpers.h"
 
 #include "revng-c/PromoteStackPointer/ComputeStackAccessesBoundsPass.h"
 #include "revng-c/PromoteStackPointer/InstrumentStackAccessesPass.h"
 #include "revng-c/Support/FunctionTags.h"
 
 using namespace llvm;
 
-bool ComputeStackAccessesBoundsPass::runOnModule(Module &M) {
+template<typename T, std::ranges::range R>
+bool areAll(const R &Range) {
+  return llvm::all_of(Range, [](auto *Object) { return isa<T>(Object); });
+}
+
+static ConstantRange
+getSCEVBoundaries(ScalarEvolution &SE, PostDominatorTree &PDT, Instruction *I) {
+  const auto *SCEV = SE.getSCEV(I);
+  auto FullSet = ConstantRange::getFull(SCEV->getType()->getIntegerBitWidth());
+
+  auto *AddRec = dyn_cast<SCEVAddRecExpr>(SCEV);
+  if (AddRec == nullptr)
+    return FullSet;
+
+  // Only consider loop ins loop simplify form
+  auto *Loop = AddRec->getLoop();
+  if (not Loop->isLoopSimplifyForm())
+    return FullSet;
+
+  // Only provide a result if the block post-dominates the header, which
+  // implies the block sees one more value than the backedge taken count
+  BasicBlock *Location = I->getParent();
+  if (not PDT.dominates(Location, Loop->getHeader()))
+    return FullSet;
+
+  // Collect operands
+  // TODO: handle multiple AddRecs
+  const auto *Start = AddRec->getStart();
+  auto *BackedgeTakenCount = SE.getBackedgeTakenCount(Loop);
+  const auto *Stride = AddRec->getStepRecurrence(SE);
+  APInt One(Start->getType()->getIntegerBitWidth(), 1);
+
+  if (not areAll<SCEVConstant>(std::vector{
+        Start, BackedgeTakenCount, Stride })) {
+    return FullSet;
+  }
+
+  // Compute end value
+  llvm::SmallVector<const llvm::SCEV *, 4> Addends = {
+    Start, SE.getMulExpr(Stride, BackedgeTakenCount), SE.getConstant(One)
+  };
+  auto *End = SE.getAddExpr(Addends);
+
+  // Ensure bounds could be computed and are constant
+  if (not areAll<SCEVConstant>(std::vector{ Start, End }))
+    return FullSet;
+
+  APInt StartValue = cast<SCEVConstant>(Start)->getAPInt();
+  APInt EndValue = cast<SCEVConstant>(End)->getAPInt();
+
+  // TODO: handle backward loops
+  if (EndValue.slt(StartValue))
+    return FullSet;
```

Comment:
Maybe I could, but I think I'm assuming FullSet means "know nothing". And if `EndValue.slt(StartValue)` it's not like a regular counted loop.

If I do this, we should probably find some threshold to say "if the range is bigger than X", bail out.
Basically I'm trying to handle only the very nice cases, but  see that's not perfect.

What do you suggest?

Reply:
Yes.

Reply:
I'm checking how `ScalarEvolution` handles overflows/underflows

Reply:
So, to do it properly you'd have to deal with `NSW` flags.

But basically you can manually check no overflow/underflow happens before calling `getAddExpr`.
If no wrap happens you can do like I suggested above, because no wrap is guaranteed.
Otherwise you return `FullSet` on wrap and you're done.

How does that sound?

Reply:
It seems to me that handling negative strides is marginally more complex, so I think it's worth it to handle it now while we're fresh on the topic.

Reply:
And I understand that you have to handle both overflow with positive stride and underflow with negative stride. But we have facilities, like `OverflowSafeInt`, that should make it easier.

Reply:
I now handle backward loops, added another discussion.